### PR TITLE
Fix for directUrl usage during testing

### PR DESCRIPTION
**Pull request by aslaker**
Fix for #7886 

This PR fixes the issue mentioned above by introducing the concept of a `TEST_DIRECT_URL` for robust test setups that need to explicitly define separate urls for their test database and their test direct database.  It also properly falls back to use the `TEST_DATABASE_URL` (and also the `cacheDirDb`) in place of the `DIRECT_URL` env variable during the test run if a `TEST_DIRECT_URL` is not set.

This will leave the current process of setting a `TEST_DATABASE_URL` unchanged, and will allow usage in situations where the user is leveraging the `directUrl` option in Prisma.  This will also prevent the Prisma commands run during the test process from overwriting data in production databases by ignoring the `TEST_DATABASE_URL` and using the value of the `directUrl`.

### CODE REVIEW

**Comment by jtoar** [comment]
Thanks @aslaker! Just discussed this with @thedavidprice, and we're worried this fix won't cover the case where `directUrl` is set to a differently-named env var like (`DIRECTURL`, etc). Is there a way we can parse the `schema.prisma` to make sure we're overwriting the right one?

Or instead of trying to overwrite the right one, we could run a quick RegExp on the `schema.prisma` and check if it's called `DIRECT_URL`, and if it's not, throw an error. Thoughts?

**Comment by aslaker** [comment]
@jtoar good point.  I had started looking into what it would take to parse the `schema.prisma` file at the beginning, but then dropped it to try out this method.  It looks like their might be a couple of libraries available to parse the file as this is something others are trying to accomplish as well.  A regex might not be a bad solution either as we wouldn't need anything so robust for this simple lookup.  I suppose a regex in combination with an update to the documentation would do the trick.

Another solution would be to see if I can put together a regex that searches the file for the `directUrl('{VALUE}')` pattern and see if I can extract the value portion and replace it during testing.  Thoughts on that?

**Comment by jtoar** [comment]
> Another solution would be to see if I can put together a regex that searches the file for the directUrl('{VALUE}') pattern and see if I can extract the value portion and replace it during testing. Thoughts on that?

This would be ideal! But throwing an error is also ok.

**Comment by aslaker** [comment]
Ok @jtoar  and @thedavidprice this should do it.  

I abstracted the logic into a function that does a regex check within the `schema.prisma` file and if there is a match it uses a second regex to extract the value of the env variable passed as the direct url.  This then gets overwritten first by `TEST_DIRECT_URL` if it exists.  If that doesn't exist, then it gets overwritten by `TEST_DATABASE_URL` and in case that doesn't exist it defaults to the `cachDirDB` located in `.redwood` or `test.db`.

The only other thing I could think of possibly doing here is considering a test for that function that checks the prisma schema (so lmk there), but otherwise we should be good!

**Comment by jtoar** [comment]
Thanks @aslaker, I'll tweak a few things but let me take it from here and get a patch out. Many thanks!

File: packages/testing/src/api/directUrl.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import path from 'path'
+
+export function getDefaultDb() {
+  return `file:${path.join(__dirname, '.redwood', 'test.db')}`
+}
```
Comment (jtoar) [inline packages/testing/src/api/directUrl.ts:5]: Ah this doesn't look quite right now that I see it again

File: packages/testing/config/jest/api/globalSetup.js [hunk 2]
```diff
@@ -25,6 +39,7 @@ module.exports = async function () {
       shell: true,
       env: {
         DATABASE_URL: process.env.DATABASE_URL,
+        DIRECT_URL: process.env.DIRECT_URL,
```
Comment (jtoar) [inline packages/testing/config/jest/api/globalSetup.js:42]: I think this part isn't taking into account the work we're doing now

**Comment by jtoar** [comment]
I think it's gtg now @aslaker if you wanted to have a second look just to make sure I didn't miss anything

File: packages/testing/config/jest/api/globalSetup.js [hunk 3]
```diff
@@ -25,6 +39,7 @@ module.exports = async function () {
       shell: true,
       env: {
         DATABASE_URL: process.env.DATABASE_URL,
+        [directUrlEnvVar]: process.env[directUrlEnvVar],
```
Comment (aslaker) [inline packages/testing/config/jest/api/globalSetup.js:42]: @jtoar good catch here.  I forgot to refactor that.  Thank you!

File: packages/testing/src/api/directUrlHelpers.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    return
+  }
+
+  // If it is, set its env var to the test equivalent.
+  const directUrlEnvMatch = directUrl[0].match(BETWEEN_PARENTHESES_REGEXP) //[2]
+
+  // This is mostly to please TS. But it's good to be safe because in this case we want to be 100% correct.
+  if (!directUrlEnvMatch) {
+    throw new Error(
+      'Error parsing `directUrl` from schema.prisma. Proceeding with this env var could be dangerous. Please check your schema.prisma file; if everything looks ok, file an issue.'
```
Comment (aslaker) [inline packages/testing/src/api/directUrlHelpers.ts:25]: 👍🏻

**Comment by aslaker** [comment]
@jtoar looks great!  Thanks so much for your help, and thanks for abstracting that logic out the proper way for me.  It's good to see how you approach that so I can apply that same standard in future PRs.

**Comment by dthyresson** [comment]
Many thanks @aslaker for this fix and contribution. It'll be a big help to those using the directUrl with Supabase alongside pgBouncer and still run migrations.

**Comment by aslaker** [comment]
@dthyresson my pleasure.  I've been wanting to use Redwood for years, and now that I'm finally using it in a project I'm very excited that I can contribute to it at the same time.

File: packages/testing/src/api/__tests__/directUrlHelpers.test.ts
Comment (thedavidprice) [inline packages/testing/src/api/__tests__/directUrlHelpers.test.ts:1]: Great work on this test 🚀

Comment (aslaker) [reply packages/testing/src/api/__tests__/directUrlHelpers.test.ts:1]: It was all @jtoar

**Pull request merged**